src/lib/Microsoft.Health.Events/EventHubProcessor/PartitionLockingBackgroundService.cs

```diff
@@ -0,0 +1,487 @@
+using Azure.Storage.Blobs.Specialized;
+using EnsureThat;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Health.Events.Common;
+using Microsoft.Health.Events.EventCheckpointing;
+using Microsoft.Health.Events.EventConsumers.Service;
+using Microsoft.Health.Events.Telemetry;
+using Microsoft.Health.Events.Telemetry.Exceptions;
+using Microsoft.Health.Logging.Telemetry;
+
+// Description
+// - First, this class gets the number of total processors and total partitions
+// - Next, it will calculate how many partitions should be assigned to each processor
+// - Next, it will lock the partition(s) using blob leases
+// - Next, it will pass the partition id(s) into the AssignedPartitionProcessor which will run until the number of total processors changes
+
+// For a flowchart of the PartitionLockingBackgroundService see:
+// \iomt-fhir\docs\PartitionLocking.md
+
+// todos:
+// - detect if a partition has no ownership updates for some time
+// - handle cases where the ratio of processors to partitions is not divided evenly
+// - detect if acquiring partitions takes too long
+// - replace Task.Delay with something more sophisticated
+
+namespace Microsoft.Health.Events.EventHubProcessor
+{
+    public class PartitionLockingBackgroundService : BackgroundService
+    {
+        private string _processorId;
+
+        private List<int> _ownedPartitions;
+
+        private BlobContainerClient _containerClient;
+
+        private int _totalProcessorsRunning = 0;
+
+        private string[] _eventHubPartitions;
+
+        private EventBatchingService _eventBatchingService;
+
+        private StorageCheckpointClient _checkpointClient;
+
+        private ITelemetryLogger _logger;
+
+        private PartitionLockingBackgroundServiceOptions _partitionLockingOptions;
+
+        private EventHubClientOptions _eventHubClientOptions;
+
+        public PartitionLockingBackgroundService(
+            PartitionLockingBackgroundServiceOptions partitionLockingOptions,
+            EventHubClientOptions eventHubClientOptions,
+            EventBatchingService eventBatchingService,
+            StorageCheckpointClient checkpointClient,
+            ITelemetryLogger logger)
+        {
+            _processorId = Guid.NewGuid().ToString(); // consider using the pod name when running in AKS
+            _ownedPartitions = new List<int>();
+            _partitionLockingOptions = partitionLockingOptions;
+            _eventHubClientOptions = eventHubClientOptions;
+            _eventBatchingService = eventBatchingService;
+            _checkpointClient = checkpointClient;
+            _logger = logger;
+        }
+
+        private void CreateContainerClient()
+        {
+            var tokenCredential = EnsureArg.IsNotNull(_partitionLockingOptions.StorageTokenCredential, nameof(_partitionLockingOptions.StorageTokenCredential));
+            _containerClient = new BlobContainerClient(_partitionLockingOptions.BlobContainerUri, tokenCredential.GetCredential());
```

Comment:
I would recommend creating an interface and wrapper class around BlobContainterClient (maybe IPartitionCoordinator and PartitionCoordinator). Currently, there is no way to Mock the BlobContainerClient which means there is no way to write unit tests for this class. Also, this class doesn't really need to know about BlobContainerClient or BlobClient it just needs a way to interact with a persistence layer in a very specific way. Abstracting away the specific storage account implementations behind a generalized interface will make testing easier and reduce the amount of code in this single class.

For example:

```c

public interface IPartitionCoordinator
{
    // Registers a processorId.
    Task ResisterProcessorIdAsync(string processorId, CancellationToken cancellationToken);

    // Returns a collection of processor ids representing active processors.
    Task<IEnumerable<string>> GetActiveProcessorIdsAsync(CancellationToken cancellationToken);

    // Checks if a partition with the given is is active (has been active within the maxInactiveTime).
    Task<bool> IsPartitionActiveAsync(string partitionId, TimeSpan maxInactiveTime, CancellationToken);

    // Claims a partition with the given partitionId.
    Task<bool> ClaimPartitionAsync(string partitionId, CancellationToken);

    ...etc...
}
```

Reply:
added new interface and class, thanks for the suggestion.